### PR TITLE
[JAX] Support FP8 training for Pipeline Parallelism when Micro-batch > 1 on Paxml.

### PR DESCRIPTION
**Pull request by mingxu1067**
- Adjust the order of updating amax history and updating scale to align with PyTorch.
- Add custom data type for FP8 meta to specify the op for gradient summation as MAX.
- Depends on [Flax/PR-3827](https://github.com/google/flax/pull/3827)

### CODE REVIEW

**Review by denera — CHANGES_REQUESTED** [review]

File: transformer_engine/jax/layernorm.py [hunk 1]
```diff
@@ -162,6 +164,12 @@ def _layernorm_fp8_dot_fwd_rule(
     k_contracting_dims = (0,)
     assert x.shape[-1] == kernel.shape[0]
 
+    is_fp8_meta_fm32 = amax.dtype == FP8MetaFP32
+    if is_fp8_meta_fm32:
+        fp8_max, amax, scale, scale_inv = \
+            convert_fp8_meta_fm32tofp32(fp8_max, amax, scale, scale_inv)
+
+    scale, scale_inv = FP8Helper.update_fp8_scale(fp8_max, amax, scale)
```
[the remark this answers is not in the document]
Comment (mingxu1067) [reply transformer_engine/jax/layernorm.py:170]: Explicitly calling the conversion here is because the conversion is for whole functions, instead of `update_fp8_scale` only. Explicitly calling the conversion ahead can make sure the dtype of FP8 meta inside the function (like `layernorm_fp8_dot`) would be the FP32, regardless the order to the computation and `update_fp8_scale`, like
```python
conversion
computation # Require FP32
update_scale # Require FP32
```

File: transformer_engine/jax/fp8.py [hunk 2]
```diff
@@ -67,6 +68,43 @@ def _format2dtypes(format_: Format):
     return jnp.bfloat16, jnp.bfloat16
 
 
+# fm32 is a custom dtype to specify the "add" rules as max operation.
+# This is typically used in Pipeline Parallelism + "MiconBatching > 1",
+# which is implemented via nn.scan. Without this custom dtype, nn.scan
+# would sum gradients from all micro-batches, and this is not the expected
+# behavior for FP8 meta. Instead, the summation of FP8 meta gradients should
+# be "MAX".
+FP8MetaFP32 = fp8_ops.fm32
```
[the remark this answers is not in the document]
Comment (mingxu1067) [reply transformer_engine/jax/fp8.py]: Good point, Done.

File: transformer_engine/jax/fp8.py [hunk 3]
```diff
@@ -67,6 +68,43 @@ def _format2dtypes(format_: Format):
...
+    assert scale_inv.dtype == FP8MetaFP32
+
+    fp8_max = jax.lax.convert_element_type(fp8_max, jnp.float32)
+    amax = jax.lax.convert_element_type(amax, jnp.float32)
+    scale = jax.lax.convert_element_type(scale, jnp.float32)
+    scale_inv = jax.lax.convert_element_type(scale_inv, jnp.float32)
+    return fp8_max, amax, scale, scale_inv
+
+
+def convert_fp8_meta_fp32tofm32(fp8_max, amax, scale, scale_inv):
```
[the remark this answers is not in the document]
Comment (mingxu1067) [reply transformer_engine/jax/fp8.py]: Done

File: transformer_engine/jax/layernorm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (denera) [reply transformer_engine/jax/layernorm.py:170]: That's a valid point. You're right, we need these separated.

**Review by denera — APPROVED** [review]
LGTM!

File: transformer_engine/jax/fp8.py [hunk 4]
```diff
@@ -303,6 +313,34 @@ def _update_fp8_metas_impl(fp8_metas: Collection) -> Collection:
...
+        assert fp8_max.dtype == FlaxFloatMeta32
+        assert amax.dtype == FlaxFloatMeta32
+        assert scale.dtype == FlaxFloatMeta32
+        assert scale_inv.dtype == FlaxFloatMeta32
+
+        fp8_max = jax.lax.convert_element_type(fp8_max, jnp.float32)
+        amax = jax.lax.convert_element_type(amax, jnp.float32)
+        scale = jax.lax.convert_element_type(scale, jnp.float32)
+        scale_inv = jax.lax.convert_element_type(scale_inv, jnp.float32)
+        return fp8_max, amax, scale, scale_inv
```
Comment (zlsh80826) [inline transformer_engine/jax/fp8.py:328]: I noticed that the `convert_fp8_meta_fm32tofp32` is always paired with `is_fp8_meta_fm32`, could we encapsulate them into a single util function as `maybe_convert_fm32_to_fp32`?

```
def may_convert_fm32_to_fp32(*args):
    return list(map(
        lambda x: jax.lax.convert_element_type(x, jnp.float32) if x.dtype == FlaxFloatMeta32 else x,
        args))
```

This can also hide `FlaxFloatMeta32` to be exposed on the other module files.

File: transformer_engine/jax/fp8.py [hunk 5]
```diff
@@ -303,6 +313,34 @@ def _update_fp8_metas_impl(fp8_metas: Collection) -> Collection:
...
+        assert fp8_max.dtype == jnp.float32
+        assert amax.dtype == jnp.float32
+        assert scale.dtype == jnp.float32
+        assert scale_inv.dtype == jnp.float32
+
+        fp8_max = jax.lax.convert_element_type(fp8_max, FlaxFloatMeta32)
+        amax = jax.lax.convert_element_type(amax, FlaxFloatMeta32)
+        scale = jax.lax.convert_element_type(scale, FlaxFloatMeta32)
+        scale_inv = jax.lax.convert_element_type(scale_inv, FlaxFloatMeta32)
+        return fp8_max, amax, scale, scale_inv
```
Comment (zlsh80826) [inline transformer_engine/jax/fp8.py:342]: same

File: transformer_engine/jax/fp8.py (hunk 4 above)
Comment (denera) [reply transformer_engine/jax/fp8.py]: I'm not sure if I understand why this would be a problem.

`convert_fp32_to_fm32()` can selectively convert only non-`fm32` inputs and return the rest unchanged, thus guaranteeing that all outputs are `fm32`.

Likewise, `convert_fm32_to_fp32()` can selectively convert only non-`fp32` inputs and return the rest unchanged, thus guaranteeing that all outputs are `fp32`.

This means we would be doing two type-checks: once in the FWD when we call `convert_fp32_to_fm32()`, and a second time in the BWD when we call `convert_fm32_to_fp32()`. Is this a significant overhead?

Comment (mingxu1067) [reply transformer_engine/jax/fp8.py]: It is about the code structure issues.
Since there are two cases in BWD to call `maybe_convert_fp32_to_fm32`, depending on if calling `convert_fm32_to_fp32`, `maybe_convert_fp32_to_fm32` could have two cases from the same input dtype. One is from fp32 to fp32 and one is fp32 to fm32. Then `maybe_convert_fp32_to_fm32` would require extra indicator to tell "should it convert fp32 to fm32 , or not`. Therefore the interface of `maybe_convert_fp32_to_fm32` would be something like:
```
def maybe_convert_fp32_to_fm32(is_converted_from_fm32, *fp8_metas)
```

Comment (denera) [reply transformer_engine/jax/fp8.py]: Okay, I think that makes sense now. The meta tensors are converted to `fm32` at the end of the backward pass only if they were originally `fm32` at the beginning of the forward pass. That would indeed require us to pass the information from the forward to the backward, so we cannot just bury the type check into the conversion function.

**Review by zlsh80826 — APPROVED** [review]

**Comment by mingxu1067** [comment]
@denera CI is passed, could you help merge this when you have any chance. Thank you

**Review by denera — APPROVED** [review]
LGTM!
